### ISSUE TITLE
Support for Win32 variable-sized array fields

### ISSUE THREAD

**Issue by n4r1b** [microsoft/windows-rs#443]
Hey @kennykerr and the whole team involved... this is Awesome!! I was really looking forward for this, great move by MS 👍 

I was wondering if there's any idea in mind on how to deal with structs that last field is a unsized arrays (ANYSIZE_ARRAY), right now the way winapi-rs is handling this by declaring an array of size 1 (ex. [CM_NOTIFY_EVENT_DATA.DeviceHandle.Data](https://docs.rs/winapi/0.3.9/winapi/um/cfgmgr32/struct.CM_NOTIFY_EVENT_DATA_DeviceHandle.html#structfield.Data)). Imho this is a bit of a constraint since we only have one field available which in the end leads to casting to thin pointers and memory transmuting operations -- specially when we receive a struct with this fields from a callback. I know this is a difficult case, and I'm pretty sure there's not straightforward way to solve it without some "unsafe" code but I would love to see/discuss your take on this 😃  

BTW, this was already discussed in the internal forums https://internals.rust-lang.org/t/ffi-types-with-unsized-arrays/5203/2

P.S.: I'm aware there's already an issue (https://github.com/microsoft/windows-rs/issues/423) that talks about arrays but I do think this case requires a separate discussion, feel free to close the issue if I'm wrong

**Comment by kennykerr** [comment]
Thanks @n4r1b! Yes, this is a tricky one and worthy of discussion. It has arrays and unions - both problematic. When the variable-sized array is trailing, what I tend to do in C++ is to omit the field entirely and simply have an enclosing struct provide the data suitable to the application, assuming the alignment and padding are compatible. I agree an array of size 1 is not very useful but is sometimes the least awful approach. Another approach is illustrated in how I modeled `windows::HString`'s  internal header:

https://github.com/microsoft/windows-rs/blob/256538e85f2f9c860c09c318423b3be79b36b9c3/src/strings/hstring.rs#L216-L224

There is a fine line between too much and too little intervention by the code generator. I will leave this open and spend some time on this and other similar structs like `OVERLAPPED` that also tend to have trailing data.

**Comment by n4r1b** [comment]
The way `windows::HString` is modeled looks very promising tbh, I've never used `MaybeUninit` but seems like a good idea to keep the unitialized instance and avoid problems with unknown size at compilation time.

My only concern with `MaybeUninit` is, and again I've never used it before so I'm not aware of the possible implications:
> Note that dropping a MaybeUninit<T> will never call T's drop code. It is your responsibility to make sure T gets dropped if it got initialized.

Would this mean that an implicit call to drop will be needed to free the memory, could this potentially lead to memory leaks?

Correct me if I'm wrong but could it be sth like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e4a3e375258e6b4b9f59578b9a4b5457

**Comment by kennykerr** [comment]
`TOKEN_PRIVILEGES` is another variant of an array field with a variable size. Just leaving that here to remind me to cover that.

**Comment by djdisodo** [comment]
will const generics fit this case?
```rust
#[repr(C)]
pub struct TOKEN_PRIVILEGES<const SIZE: usize> {
    pub PrivilegeCount: u32,
    pub Privileges: [LUID_AND_ATTRIBUTES; SIZE],
}
```

**Comment by kennykerr** [comment]
Closing as duplicate of #907 which is blocked on https://github.com/microsoft/win32metadata/issues/547
